### PR TITLE
fix: inconsistent image normalization for wandb.Image

### LINKED ISSUE #6287: [CLI]: wandb.Image gives blanks for large or small values
### Describe the bug

According to the [wandb.Image documentation](https://docs.wandb.ai/ref/python/data-types/image), the type accepts "numpy array of image data", where it "attempts to infer the data format and converts it". This works for float-valued raw data, but not if the values are very large or very small in scale. [The code doing the conversion](https://github.com/wandb/wandb/blob/main/wandb/sdk/data_types/image.py#L269) has different pathways for numpy arrays and tensors. Torch tensors work properly for large values, but not for small values.

If an array of values of scale 1e-5 shows as white-to-black noise, it doesn't make much sense that an array of 1e-8 values would show as full-black.

```python
>>> import numpy as np
>>> import wandb
>>> wandb.Image(np.random.uniform(size=[100, 200]) * 1e-8).image.save("preview_e-8.png")
>>> wandb.Image(np.random.uniform(size=[100, 200])).image.save("preview_e0.png")
>>> wandb.Image(np.random.uniform(size=[100, 200]) * 1e8).image.save("preview_e8.png")
>>> wandb.Image(torch.rand(size=[100, 200]) * 1e-8).image.save("torch_preview_e-8.png")
>>> wandb.Image(torch.rand(size=[100, 200])* 1e-5).image.save("preview_e-5.png")
>>> wandb.Image(torch.rand(size=[100, 200]) * 1e8).image.save("torch_preview_e8.png")
```

No exception, but the images look very different even though they should look qualitatively similar. `preview_e-8.png` is a fully black image, `preview_e0.png` is as expected (noise), and `preview_e8.png` is a fully white image. `preview_e-5.png` is the first scale where we get noise, at e-6 we don't get black but the white values become dark gray. The conversion works for large torch tensors, and `torch_preview_e8.png` shows noise, but `torch_preview_e-8.png` is a black image.

### Additional Files

_No response_

### Environment

WandB version: 0.15.0

OS: macOS 13.0 Ventura

Python version: 3.10.10

Versions of relevant libraries:

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by jacobromero**
Description
-----------

- Fixes WB-15801
- Fixes #6287, #7377

What does the PR do? Include a concise description of the PR contents.
When normalizing image values, during the creation of a `wandb.Image` from data (numpy array, or pytorch tensor) there are inconsistencies in the output image. 
For some small values normalized and converted image will be entirely black, or for small negative values the image will be entirely white. Additionally, between passing in a numpy array or a pytorch tensor would also yield different image results.

This PR uses the same normalization between the image data passed into the `wandb.Image` initialization function. As well as adds a flag to allow a user to normalize image data, since this can affect the output results of the image which might be undesirable for some users.

- [ ] I updated CHANGELOG.unreleased.md, or it's not applicable

Testing
-------
How was this PR tested?
- Unit tests added
- repro script
```python
import numpy as np
import wandb
import torch

img = np.random.uniform(low=0, high=1, size=[2, 4, 3])
torch_img = torch.from_numpy(img.transpose(2, 0, 1))
values = [1e-8, 1e-5, 1e0, 1e1]

with wandb.init(
    project="image-norm-test",
) as run:
    for value in values:
        wandb_np_img_norm = wandb.Image(img * value)
        wandb_torch_img_norm = wandb.Image(torch_img * value)

        run.log({f"np_img_norm_{str(value)}": wandb_np_img_norm})
        run.log({f"torch_img_norm_{str(value)}": wandb_torch_img_norm})
```

#### After this change
image
image
image
image

#### Before this change
image
image
image
image

### CODE REVIEW

File: wandb/sdk/data_types/image.py [hunk 1]
```diff
@@ -309,17 +326,19 @@ def _initialize_from_data(
...
             if hasattr(data, "requires_grad") and data.requires_grad:
                 data = data.detach()  # type: ignore
             if hasattr(data, "dtype") and str(data.dtype) == "torch.uint8":
                 data = data.to(float)
-            data = vis_util.make_grid(data, normalize=True)
             mode = mode or self.guess_mode(data, file_type)
+            data = data.permute(1, 2, 0).cpu().numpy()
+            data = self.normalize_and_convert_to_uint8(data) if normalize else data
+
+            if data.ndim > 2:
```
Comment (dmitryduev) [inline wandb/sdk/data_types/image.py:337]: why? would this potentially squeeze out an alpha dim?

Comment (jacobromero) [reply wandb/sdk/data_types/image.py:337]: Nope this was being done before as well.

Basically be because pytorch has [the convention](https://discuss.pytorch.org/t/dimensions-of-an-input-image/19439/2) for images to be `(# channels, height, width)` and we expect numpy arrays to be `(height, width, # channels)`, all we are doing re-arranging the pytorch tensor to be in the same format as the numpy array.

**Review by dmitryduev — APPROVED** [review]

File: wandb/sdk/data_types/image.py [hunk 2]
```diff
@@ -505,28 +525,17 @@ def guess_mode(
...
-        dmin = np.min(data)
-        if dmin < 0:
-            data = (data - np.min(data)) / np.ptp(data)
-        if np.max(data) <= 1.0:
-            data = (data * 255).astype(np.int32)
-
-        # assert issubclass(data.dtype.type, np.integer), 'Illegal image format.'
+        # Scale data to [0, 255] range
+        data = data - np.min(data)
+        data = data / np.ptp(data)
```
Comment (timoffex) [inline wandb/sdk/data_types/image.py:537]: If `np.ptp(data)` is zero, then isn't this 0/0?

Comment (timoffex) [reply wandb/sdk/data_types/image.py:537]: Does `np.clip()` fix that, maybe? But would it replace everything by 0 or 255?

File: wandb/sdk/data_types/image.py [hunk 3]
```diff
@@ -151,7 +93,81 @@ def __init__(
         boxes: Optional[Union[Dict[str, "BoundingBoxes2D"], Dict[str, dict]]] = None,
         masks: Optional[Union[Dict[str, "ImageMask"], Dict[str, dict]]] = None,
         file_type: Optional[str] = None,
+        normalize: bool = True,
```
Comment (timoffex) [inline wandb/sdk/data_types/image.py:96]: Before this PR, most images would be unaffected, but now, all images that aren't a constant value have a pure black and a pure white pixel. I don't think True is a good default for this option; in fact, pre-processing the data to a desired format is something the user should do, and we just need to tell them the expectation (should values be integers in the range 0..255, or floats in the range 0..1?)

Comment (jacobromero) [reply wandb/sdk/data_types/image.py:96]: I agree that we shouldnt really be operating on user data, especially before we were manipulating the data without telling the users.

Given that we are planning a 0.20 release next, we can remove normalization all together, or default to `false`.
The idea of keeping it to true, was that we are already modifying the user provided data.
